services/src/main/java/com/chatbot/services/ChatServiceController.java

```diff
@@ -0,0 +1,185 @@
+      "https://www.googleapis.com/service_accounts/v1/metadata/x509/";
+  private static final String AUDIENCE = System.getenv("projectNumber");
+
+  @PostMapping("/")
+  public Message onRequest(@RequestHeader final Map<String, String> headers,
+      @RequestBody final JsonNode event) throws IOException, GeneralSecurityException {
+    if (headers.get("user-agent").equals(HANGOUTS_USER_AGENT)) {
+      final String BEARER_TOKEN = headers.get("authorization").split(" ")[1];
+      final JacksonFactory factory = new JacksonFactory();
+      final GooglePublicKeysManager.Builder keyManagerBuilder = new GooglePublicKeysManager.Builder(
+          new NetHttpTransport(), factory);
+      final String certUrl = PUBLIC_CERT_URL_PREFIX + CHAT_ISSUER;
+      keyManagerBuilder.setPublicCertsEncodedUrl(certUrl);
+      final GoogleIdTokenVerifier.Builder verifierBuilder = new GoogleIdTokenVerifier.Builder(
+          keyManagerBuilder.build());
+      verifierBuilder.setIssuer(CHAT_ISSUER);
+      final GoogleIdTokenVerifier verifier = verifierBuilder.build();
+      final GoogleIdToken idToken = GoogleIdToken.parse(factory, BEARER_TOKEN);
+      if (idToken == null) {
+        throw new IllegalArgumentException("Token cannot be parsed");
+      }
+      if (!verifier.verify(idToken) || !idToken.verifyAudience(Collections.singletonList(AUDIENCE))
+          || !idToken.verifyIssuer(CHAT_ISSUER)) {
+        throw new IllegalArgumentException("Invalid token");
+      }
+      try {
+        asyncService.hangoutsAsyncHandler(parseHangoutsRequest(event));
+      } catch (final Exception e) {
+        // If there was an error in parsing the request, either we do not support the
+        // type of request or the format of the request is incorrect, in both these cases
+        // returning an empty string is an option.
+        e.printStackTrace();
+      }
+    } else {
+      // parseWhatsappRequest(event);
+      // whatsapp async handler
+      // return acknowledgement
+    }
+    final Message reply = new Message();
+    if (!event.at("/type").asText().equals(CARD_CLICKED_EVENT)) {
+      // an empty message acts as an acknowledgement for a text message
+      reply.setText("");
+    } else {
+      // replace the button with a text paragraph to render it unclickable
+      reply.setActionResponse((new ActionResponse()).set("type", "UPDATE_MESSAGE"));
+      final List<WidgetMarkup> widgets = new ArrayList<>();
+      final TextParagraph widget =
+          new TextParagraph().setText(event.at("/action/parameters/0/value").asText());
+      widgets.add(new WidgetMarkup().setTextParagraph(widget));
+      final Section section = new Section()
+          .setWidgets(widgets);
+      reply.setCards(Collections
+          .singletonList((new Card()).setSections(Collections.singletonList(section))));
+    }
+    return reply;
+  }
+
+  private ChatServiceRequest parseHangoutsRequest(final JsonNode event) throws Exception {
```

Comment:
you are also building a chat service request here. Change the function name to something like BuildChatServiceRequestFromHangoutsRequest

Reply:
updated